custom-elements/reactions/DOMTokenList.html

```diff
@@ -0,0 +1,223 @@
+    instance.classList = 'hello';
+    var logEntries = element.log();
+    assert_array_equals(logEntries.types(), ['attributeChanged']);
+    assert_equals(logEntries.log().name, 'class');
+    assert_equals(logEntries.log().oldValue, null);
+    assert_equals(logEntries.log().newValue, 'hello');
+    assert_equals(logEntries.log().namespace, null);
+}, 'the stringifier of DOMTokenList must enqueue a attributeChanged reaction when adding an observed attribute');
+
+test(function () {
+    var element = defineNewCustomElement([]);
+    var instance = document.createElement(element.name);
+    instance.setAttribute('class', 'hello');
+    assert_array_equals(element.log().types(), ['constructed']);
+    instance.classList = 'hello';
+    var logEntries = element.log();
+    assert_array_equals(element.log().types(), []);
+}, 'the stringifier of DOMTokenList must not enqueue a attributeChanged reaction when adding an unobserved attribute');
+
+test(function () {
+    var element = defineNewCustomElement(['class']);
+    var instance = document.createElement(element.name);
+    instance.setAttribute('class', 'hello');
+    assert_array_equals(element.log().types(), ['constructed', 'attributeChanged']);
+    instance.classList = 'world';
+    var logEntries = element.log();
+    assert_array_equals(logEntries.types(), ['attributeChanged']);
+    assert_equals(logEntries.log().name, 'class');
+    assert_equals(logEntries.log().oldValue, 'hello');
+    assert_equals(logEntries.log().newValue, 'world');
+    assert_equals(logEntries.log().namespace, null);
+}, 'the stringifier of DOMTokenList must enqueue a attributeChanged reaction when mutating the value of an observed attribute');
+
+test(function () {
+    var element = defineNewCustomElement([]);
+    var instance = document.createElement(element.name);
+    instance.setAttribute('class', 'hello');
+    assert_array_equals(element.log().types(), ['constructed']);
+    instance.classList = 'world';
+    assert_array_equals(element.log().types(), []);
+}, 'the stringifier of DOMTokenList must not enqueue a attributeChanged reaction when mutating the value of an unobserved attribute');
+
+test(function () {
+    var element = defineNewCustomElement(['class']);
+    var instance = document.createElement(element.name);
+    instance.setAttribute('class', 'hello');
+    assert_array_equals(element.log().types(), ['constructed', 'attributeChanged']);
+    instance.classList = 'hello';
+    var logEntries = element.log();
+    assert_array_equals(logEntries.types(), ['attributeChanged']);
+    assert_equals(logEntries.log().name, 'class');
+    assert_equals(logEntries.log().oldValue, 'hello');
+    assert_equals(logEntries.log().newValue, 'hello');
+    assert_equals(logEntries.log().namespace, null);
+}, 'the stringifier of DOMTokenList must enqueue a attributeChanged reaction when the setter is called with the original value of the attribute');
+
+</script>
```

Comment:
I wonder if we also need tests for eg add('foo', 'bar') that check we only run "update steps" once and hence only one attributeChanged callback?

Another thing to maybe test is an element with a mix of observed and unobserved attributes to ensure the implementation is making the decision about firing attributeChanged on a per-attribute and not per-element basis.